File: server/services/transaction/internal/domain/model/recurring_transaction.go
package model

import (
	"time"

	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

type RecurrenceFrequency string

const (
	FrequencyDaily   RecurrenceFrequency = "daily"
	FrequencyWeekly  RecurrenceFrequency = "weekly"
	FrequencyMonthly RecurrenceFrequency = "monthly"
	FrequencyYearly  RecurrenceFrequency = "yearly"
)

func IsValidFrequency(f RecurrenceFrequency) bool {
	switch f {
	case FrequencyDaily, FrequencyWeekly, FrequencyMonthly, FrequencyYearly:
		return true
	}
	return false
}

type RecurringTransaction struct {
	ID          uint                `gorm:"primaryKey" json:"id"`
	UserID      uint                `gorm:"index;not null" json:"user_id"`
	AccountID   uint                `gorm:"index;not null" json:"account_id"`
	Account     *Account            `gorm:"foreignKey:AccountID" json:"-"`
	Type        TransactionType     `gorm:"type:varchar(20);not null" json:"type"`
	Amount      decimal.Decimal     `gorm:"type:decimal(19,4);not null" json:"amount"`
	Currency    string              `gorm:"type:char(3);default:'RUB'" json:"currency"`
	CategoryID  *uint               `gorm:"index" json:"category_id,omitempty"`
	Category    *Category           `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
	Description string              `gorm:"type:text" json:"description"`
	Frequency   RecurrenceFrequency `gorm:"type:varchar(20);not null" json:"frequency"`
	NextDate    time.Time           `gorm:"not null" json:"next_date"`
	EndDate     *time.Time          `json:"end_date,omitempty"`
	IsActive    bool                `gorm:"default:true" json:"is_active"`
	CreatedAt   time.Time           `json:"created_at"`
	UpdatedAt   time.Time           `json:"updated_at"`
	DeletedAt   gorm.DeletedAt      `gorm:"index" json:"-"`
}

File: app/android/app/src/main/java/com/barghest/bux/domain/model/RecurringTransaction.kt
package com.barghest.bux.domain.model

import java.math.BigDecimal
import java.time.Instant

data class RecurringTransaction(
    val id: Int,
    val accountId: Int,
    val type: TransactionType,
    val amount: BigDecimal,
    val currency: String,
    val categoryId: Int?,
    val description: String?,
    val frequency: RecurrenceFrequency,
    val nextDate: Instant,
    val endDate: Instant?,
    val isActive: Boolean
)

data class NewRecurringTransaction(
    val accountId: Int,
    val type: TransactionType,
    val amount: BigDecimal,
    val currency: String,
    val categoryId: Int? = null,
    val description: String? = null,
    val frequency: RecurrenceFrequency,
    val nextDate: Instant,
    val endDate: Instant? = null
)

enum class RecurrenceFrequency(val value: String, val label: String) {
    DAILY("daily", "Ежедневно"),
    WEEKLY("weekly", "Еженедельно"),
    MONTHLY("monthly", "Ежемесячно"),
    YEARLY("yearly", "Ежегодно");

    companion object {
        fun fromValue(value: String): RecurrenceFrequency {
            return entries.find { it.value == value } ?: MONTHLY
        }
    }
}
